File: scripts/geometry_asset_version.py
"""Content-derived cache token shared by Python page generators."""
from __future__ import annotations

import hashlib
import json
from pathlib import Path

GEOMETRY_ASSETS = (
    "data/geometry-route-contracts.json",
    "css/alive.css",
    "js/mandala.js",
    "js/indra.js",
)


def geometry_exemption_for_path(contracts: dict, relative_html_path: str) -> str | None:
    coverage = contracts.get("coverage", {})
    star_routes = coverage.get("star_page_routes")
    control_prefixes = coverage.get("control_page_prefixes")
    if (
        not isinstance(star_routes, list)
        or len(star_routes) != int(coverage.get("expected_current_star_pages", -1))
        or len(set(star_routes)) != len(star_routes)
    ):
        raise ValueError("geometry contract must declare the exact canonical star_page_routes")
    if not isinstance(control_prefixes, list) or not control_prefixes:
        raise ValueError("geometry contract must declare source-only control_page_prefixes")
    relative = str(relative_html_path).replace("\\", "/").lstrip("/")
    if relative in star_routes:
        return str(coverage.get("star_page_exemption_value") or "star-file")
    if any(relative.startswith(str(prefix)) for prefix in control_prefixes):
        return str(coverage.get("control_page_exemption_value") or "internal-control")
    return None


def is_star_page(contracts: dict, relative_html_path: str) -> bool:
    return geometry_exemption_for_path(contracts, relative_html_path) == contracts.get(
        "coverage", {}
    ).get("star_page_exemption_value")


def is_geometry_exempt(contracts: dict, relative_html_path: str) -> bool:
    return geometry_exemption_for_path(contracts, relative_html_path) is not None


def opacity_bounds(contracts: dict) -> tuple[float, float]:
    bounds = contracts.get("presentation", {}).get("opacity_bounds", {})
    minimum = float(bounds.get("minimum", -1))
    maximum = float(bounds.get("maximum", -1))
    if minimum < 0 or maximum > 1 or minimum > maximum:
        raise ValueError("geometry contract must declare valid presentation.opacity_bounds")
    return minimum, maximum


def valid_opacity(contracts: dict, value: object) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    minimum, maximum = opacity_bounds(contracts)
    return number if minimum <= number <= maximum else None


def owner_opacity_for_key(contracts: dict, key: str, route_type: str) -> float | None:
    overrides = contracts.get("presentation", {}).get("owner_fade_overrides")
    if not isinstance(overrides, list):
        raise ValueError("geometry contract must declare presentation.owner_fade_overrides")
    for override in overrides:
        selectors = [
            name
            for name in ("exact_path", "path_prefix", "route_type")
            if isinstance(override.get(name), str)
        ]
        if len(selectors) != 1:
            raise ValueError("each geometry owner fade override must have exactly one selector")
        matches = (
            override.get("exact_path") == key
            or (
                isinstance(override.get("path_prefix"), str)
                and key.startswith(override["path_prefix"])
            )
            or override.get("route_type") == route_type
        )
        if matches:
            opacity = valid_opacity(contracts, override.get("opacity"))
            if opacity is None:
                raise ValueError(
                    f"invalid geometry owner opacity for {override.get('owner', selectors[0])}"
                )
            return opacity
    return None


def geometry_asset_version(root: Path) -> str:
    root = Path(root)
    contracts = json.loads(
        (root / "data" / "geometry-route-contracts.json").read_text(encoding="utf-8")
    )
    if contracts.get("asset_version_scheme") != "sha256-12":
        raise ValueError("geometry asset_version_scheme must be sha256-12")
    digest = hashlib.sha256()
    for relative in GEOMETRY_ASSETS:
        digest.update(relative.encode("utf-8"))
        digest.update(b"\0")
        digest.update((root / relative).read_bytes())
        digest.update(b"\0")
    return f"sha256-{digest.hexdigest()[:12]}"


def geometry_body_attributes(
    root: Path,
    relative_html_path: str,
    route_type: str,
    *,
    register: str,
    profile: str = "dual-field",
    intensity: float | None = None,
) -> str:
    """Return the finalizer-compatible body contract for a generated page.

    Generators that hash one generated HTML page as the translation source for
    another must emit the final geometry attributes before taking that hash.
    The catch-all finalizer still owns every page and verifies these bytes.
    """
    root = Path(root)
    contracts = json.loads(
        (root / "data" / "geometry-route-contracts.json").read_text(encoding="utf-8")
    )
    exemption = geometry_exemption_for_path(contracts, relative_html_path)
    if exemption:
        raise ValueError(
            f"{relative_html_path}: {exemption} pages must not receive public rainbow geometry"
        )
    roles = contracts.get("route_types", {}).get(route_type)
    register_contract = contracts.get("registers", {}).get(register)
    profile_contract = contracts.get("profiles", {}).get(profile)
    if not isinstance(roles, list) or not roles:
        raise ValueError(f"unknown or empty geometry route type: {route_type}")
    if not isinstance(register_contract, dict):
        raise ValueError(f"unknown geometry register: {register}")
    if not isinstance(profile_contract, dict):
        raise ValueError(f"unknown geometry profile: {profile}")

    relative = str(relative_html_path).replace("\\", "/").lstrip("/")
    if relative == "index.html":
        key = "/"
    elif relative.endswith("/index.html"):
        key = f"/{relative[:-len('index.html')]}"
    else:
        key = f"/{relative}"
    seed = hashlib.sha256(key.encode("utf-8")).hexdigest()[:16]
    register_intensity = valid_opacity(contracts, register_contract.get("default_intensity"))
    if register_intensity is None:
        raise ValueError(f"invalid geometry register intensity: {register}")
    owner_intensity = owner_opacity_for_key(contracts, key, route_type)
    page_intensity = valid_opacity(contracts, intensity)
    resolved_intensity = (
        owner_intensity
        if owner_intensity is not None
        else page_intensity if page_intensity is not None else register_intensity
    )
    fade_source = (
        "page"
        if owner_intensity is not None or page_intensity is not None
        else "route-register"
    )
    return (
        f'data-route-type="{route_type}"'
        ' data-geometry="indra-web"'
        f' data-indra-intensity="{resolved_intensity:.3f}"'
        f' data-indra-fade-source="{fade_source}"'
        f' data-geometry-role="{" ".join(roles)}"'
        f' data-geometry-key="{key}"'
        f' data-geometry-seed="{seed}"'
        f' data-geometry-register="{register}"'
        f' data-geometry-profile="{profile}"'
        ' data-front-facing="general-audience"'
    )

File: scripts/lib/geometry-asset-version.js
'use strict';

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const GEOMETRY_ASSETS = Object.freeze([
  'data/geometry-route-contracts.json',
  'css/alive.css',
  'js/mandala.js',
  'js/indra.js',
]);
const VERSION_SCHEME = 'sha256-12';

/**
 * Return a cache token that changes whenever any shipped geometry asset does.
 * Path separators and NUL delimiters make the combined digest unambiguous.
 */
function geometryAssetVersion(root) {
  const hash = crypto.createHash('sha256');
  for (const relative of GEOMETRY_ASSETS) {
    hash.update(relative, 'utf8');
    hash.update('\0', 'utf8');
    hash.update(fs.readFileSync(path.join(root, relative)));
    hash.update('\0', 'utf8');
  }
  return `sha256-${hash.digest('hex').slice(0, 12)}`;
}

function assertGeometryVersionScheme(contracts) {
  if (!contracts || contracts.asset_version_scheme !== VERSION_SCHEME) {
    throw new Error(`geometry contract asset_version_scheme must be ${VERSION_SCHEME}`);
  }
  const coverage = contracts.coverage || {};
  const starRoutes = coverage.star_page_routes;
  const controlPrefixes = coverage.control_page_prefixes;
  if (!Array.isArray(starRoutes)
      || starRoutes.length !== Number(coverage.expected_current_star_pages)
      || new Set(starRoutes).size !== starRoutes.length) {
    throw new Error('geometry contract must declare the exact canonical star_page_routes');
  }
  if (!Array.isArray(controlPrefixes) || controlPrefixes.length === 0) {
    throw new Error('geometry contract must declare source-only control_page_prefixes');
  }
  if (coverage.exemption_attribute !== 'data-shared-geometry-exempt') {
    throw new Error('geometry contract exemption_attribute must be data-shared-geometry-exempt');
  }
  const presentation = contracts.presentation || {};
  const palette = presentation.rainbow_palette;
  if (!Array.isArray(palette) || palette.length !== 8
      || palette.some(value => !/^#[0-9a-f]{6}$/.test(String(value)))) {
    throw new Error('geometry contract must declare one immutable eight-colour lowercase palette');
  }
  geometryOpacityBounds(contracts);
}

/**
 * Return the explicit shared-geometry exemption for an HTML source path.
 * Star files are exact routes: a future ordinary project created beneath one of
 * their parent directories must not become exempt by accident. Dashboard is a
 * source-only internal control family and remains prefix-classified so any
 * additional private control views fail the expected-count gate until reviewed.
 */
function geometryExemptionForRelativeHtmlPath(contracts, relativeHtmlPath) {
  assertGeometryVersionScheme(contracts);
  const coverage = contracts.coverage;
  const relative = String(relativeHtmlPath).replace(/\\/g, '/').replace(/^\/+/, '');
  if (coverage.star_page_routes.includes(relative)) return coverage.star_page_exemption_value;
  if (coverage.control_page_prefixes.some(prefix => relative.startsWith(String(prefix)))) {
    return coverage.control_page_exemption_value;
  }
  return null;
}

function isStarPageRelativeHtmlPath(contracts, relativeHtmlPath) {
  return geometryExemptionForRelativeHtmlPath(contracts, relativeHtmlPath)
    === contracts.coverage.star_page_exemption_value;
}

function isGeometryExemptRelativeHtmlPath(contracts, relativeHtmlPath) {
  return geometryExemptionForRelativeHtmlPath(contracts, relativeHtmlPath) !== null;
}

function geometryOpacityBounds(contracts) {
  const bounds = contracts && contracts.presentation && contracts.presentation.opacity_bounds;
  const minimum = Number(bounds && bounds.minimum);
  const maximum = Number(bounds && bounds.maximum);
  if (!Number.isFinite(minimum) || !Number.isFinite(maximum)
      || minimum < 0 || maximum > 1 || minimum > maximum) {
    throw new Error('geometry contract must declare valid presentation.opacity_bounds');
  }
  return Object.freeze({ minimum, maximum });
}

function validGeometryOpacity(contracts, value) {
  const number = Number(value);
  const bounds = geometryOpacityBounds(contracts);
  return Number.isFinite(number) && number >= bounds.minimum && number <= bounds.maximum
    ? number
    : null;
}

function geometryOwnerOpacityForKey(contracts, key, routeType) {
  const overrides = contracts.presentation && contracts.presentation.owner_fade_overrides;
  if (!Array.isArray(overrides)) {
    throw new Error('geometry contract must declare presentation.owner_fade_overrides');
  }
  for (const override of overrides) {
    const selectors = ['exact_path', 'path_prefix', 'route_type']
      .filter(name => typeof override[name] === 'string');
    if (selectors.length !== 1) {
      throw new Error('each geometry owner fade override must have exactly one selector');
    }
    const matches = override.exact_path === key
      || (typeof override.path_prefix === 'string' && key.startsWith(override.path_prefix))
      || override.route_type === routeType;
    if (matches) {
      const opacity = validGeometryOpacity(contracts, override.opacity);
      if (opacity === null) throw new Error(`invalid geometry owner opacity for ${override.owner || selectors[0]}`);
      return opacity;
    }
  }
  return null;
}

/**
 * Normalize a source HTML path to the pathname-only geometry camera key used
 * by both the catch-all finalizer and page generators.
 */
function geometryKeyForRelativeHtmlPath(relativeHtmlPath) {
  const relative = String(relativeHtmlPath).replace(/\\/g, '/').replace(/^\/+/, '');
  if (relative === 'index.html') return '/';
  if (relative.endsWith('/index.html')) return `/${relative.slice(0, -'index.html'.length)}`;
  return `/${relative}`;
}

function geometryRegisterForKey(key) {
  if (/^\/polymyth\/coherence(?:\/|$)/.test(key)) return 'quiet';
  if (key === '/' || /^\/(?:polymyth|bb|bookwormcard|campaigns|aa|ohm-dome|agora|florilegium)(?:\/|$)/.test(key)) return 'expressive';
  if (/^\/(?:about|leizu|saul|teacherresources|polymythseminars|seminars|writingclub|writingkids|writingjuniors|writingteens|writinggrads|university|philosophy|humanities|cfps|lectures|fellowships|marginalia|reviews)(?:\/|$)/.test(key)) return 'quiet';
  return 'standard';
}

function geometryProfileFor(relativeHtmlPath, routeType) {
  return 'dual-field';
}

function geometryMotionPresetFor(contracts, key, routeType) {
  for (const rule of contracts.motion_preset_rules || []) {
    if (rule.exact_path === key || (rule.path_prefixes || []).some(prefix => key.startsWith(prefix))
        || (rule.route_types || []).includes(routeType)) {
      if (!contracts.motion_presets[rule.preset]) throw new Error('Unknown geometry motion preset ' + rule.preset);
      return rule.preset;
    }
  }
  return 'flow';
}

function geometrySeedForKey(key) {
  return crypto.createHash('sha256').update(key).digest('hex').slice(0, 16);
}

/**
 * Emit the canonical body contract in the exact order owned by the finalizer.
 * Keeping this in one JS helper makes generated pages byte-idempotent before
 * the finalizer runs, while the finalizer still validates every HTML page.
 */
function geometryBodyAttributes(contracts, relativeHtmlPath, routeType, options = {}) {
  assertGeometryVersionScheme(contracts);
  const exemption = geometryExemptionForRelativeHtmlPath(contracts, relativeHtmlPath);
  if (exemption) {
    throw new Error(`${relativeHtmlPath}: ${exemption} pages must not receive public rainbow geometry`);
  }
  const roles = contracts.route_types && contracts.route_types[routeType];
  if (!Array.isArray(roles) || roles.length === 0) {
    throw new Error(`${routeType}: missing structural geometry roles`);
  }
  const key = geometryKeyForRelativeHtmlPath(relativeHtmlPath);
  const register = geometryRegisterForKey(key);
  const profile = geometryProfileFor(relativeHtmlPath, routeType);
  const registerContract = contracts.registers && contracts.registers[register];
  const profileContract = contracts.profiles && contracts.profiles[profile];
  if (!registerContract || !Number.isFinite(Number(registerContract.default_intensity))) {
    throw new Error(`${register}: missing geometry register contract`);
  }
  if (!profileContract) throw new Error(`${profile}: missing geometry profile contract`);
  const registerIntensity = validGeometryOpacity(contracts, registerContract.default_intensity);
  if (registerIntensity === null) throw new Error(`${register}: invalid geometry register intensity`);
  const normalizedOptions = typeof options === 'number'
    ? { pageOwnedIntensity: options }
    : (options || {});
  const ownerOverride = geometryOwnerOpacityForKey(contracts, key, routeType);
  const pageOwned = validGeometryOpacity(contracts, normalizedOptions.pageOwnedIntensity);
  const resolvedIntensity = ownerOverride !== null
    ? ownerOverride
    : pageOwned !== null ? pageOwned : registerIntensity;
  const fadeSource = ownerOverride !== null || pageOwned !== null ? 'page' : 'route-register';
  const intensity = resolvedIntensity.toFixed(3);
  return `data-route-type="${routeType}"`
    + ' data-geometry="indra-web"'
    + ` data-indra-intensity="${intensity}"`
    + ` data-indra-fade-source="${fadeSource}"`
    + ` data-geometry-role="${roles.join(' ')}"`
    + ` data-geometry-key="${key}"`
    + ` data-geometry-seed="${geometrySeedForKey(key)}"`
    + ` data-geometry-register="${register}"`
    + ` data-geometry-profile="${profile}"`
    + ` data-geometry-motion-preset="${geometryMotionPresetFor(contracts, key, routeType)}"`
    + ' data-front-facing="general-audience"';
}

module.exports = {
  GEOMETRY_ASSETS,
  VERSION_SCHEME,
  assertGeometryVersionScheme,
  geometryBodyAttributes,
  geometryAssetVersion,
  geometryExemptionForRelativeHtmlPath,
  geometryKeyForRelativeHtmlPath,
  geometryOpacityBounds,
  geometryOwnerOpacityForKey,
  geometryProfileFor,
  geometryMotionPresetFor,
  geometryRegisterForKey,
  geometrySeedForKey,
  isGeometryExemptRelativeHtmlPath,
  isStarPageRelativeHtmlPath,
  validGeometryOpacity,
};
